File: android/app/src/androidTest/kotlin/com/menumaker/pageobjects/SignupPage.kt
package com.menumaker.pageobjects

import androidx.compose.ui.test.*
import androidx.compose.ui.test.junit4.ComposeTestRule

/**
 * Page Object for Signup Screen
 */
class SignupPage(private val composeTestRule: ComposeTestRule) {

    // Elements
    private val nameField = composeTestRule.onNodeWithText("Name", substring = true)
    private val emailField = composeTestRule.onNodeWithText("Email", substring = true)
    private val passwordField = composeTestRule.onAllNodesWithText("Password", substring = true).onFirst()
    private val confirmPasswordField = composeTestRule.onNodeWithText("Confirm Password", substring = true)
    private val phoneField = composeTestRule.onNodeWithText("Phone", substring = true)
    private val signupButton = composeTestRule.onNodeWithText("Sign Up", ignoreCase = true)
    private val loginButton = composeTestRule.onNodeWithText("Log In", ignoreCase = true)
    private val termsCheckbox = composeTestRule.onNode(hasClickAction().and(hasText("terms", substring = true, ignoreCase = true)))

    // Actions
    fun enterName(name: String): SignupPage {
        nameField.performTextInput(name)
        return this
    }

    fun enterEmail(email: String): SignupPage {
        emailField.performTextInput(email)
        return this
    }

    fun enterPassword(password: String): SignupPage {
        passwordField.performTextInput(password)
        return this
    }

    fun enterConfirmPassword(password: String): SignupPage {
        confirmPasswordField.performTextInput(password)
        return this
    }

    fun enterPhone(phone: String): SignupPage {
        phoneField.performTextInput(phone)
        return this
    }

    fun acceptTerms(): SignupPage {
        termsCheckbox.performClick()
        return this
    }

    fun tapSignup(): SignupPage {
        signupButton.performClick()
        Thread.sleep(2000)
        return this
    }

    fun tapLogin(): SignupPage {
        loginButton.performClick()
        Thread.sleep(1000)
        return this
    }

    fun signup(name: String, email: String, password: String, phone: String): SignupPage {
        enterName(name)
        enterEmail(email)
        enterPassword(password)
        enterConfirmPassword(password)
        enterPhone(phone)
        acceptTerms()
        tapSignup()
        return this
    }

    // Assertions
    fun assertScreenDisplayed(): SignupPage {
        signupButton.assertExists()
        emailField.assertExists()
        return this
    }

    fun assertSignupButtonEnabled(): SignupPage {
        signupButton.assertIsEnabled()
        return this
    }
}

File: ios/MenuMakerUITests/PageObjects/SignupPage.swift
//
//  SignupPage.swift
//  MenuMakerUITests
//
//  Page Object for Signup Screen
//

import XCTest

struct SignupPage {
    let app: XCUIApplication

    // MARK: - Elements

    private var signupScreen: XCUIElement {
        app.scrollViews["signup-screen"]
    }

    var nameField: XCUIElement {
        signupScreen.textFields["name-field"]
    }

    var emailField: XCUIElement {
        signupScreen.textFields["email-field"]
    }

    var phoneField: XCUIElement {
        signupScreen.textFields["phone-field"]
    }

    var passwordField: XCUIElement {
        signupScreen.secureTextFields["password-field"]
    }

    var confirmPasswordField: XCUIElement {
        signupScreen.secureTextFields["confirm-password-field"]
    }

    var signupButton: XCUIElement {
        app.buttons["signup-button"]
    }

    var loginLink: XCUIElement {
        app.buttons["cancel-button"]
    }

    var errorMessage: XCUIElement {
        signupScreen.staticTexts["error-message"]
    }

    var validationMessage: XCUIElement {
        // Validation messages use the same error-message identifier
        signupScreen.staticTexts["error-message"]
    }

    // MARK: - Actions

    @discardableResult
    func enterName(_ name: String) -> SignupPage {
        nameField.tap()
        nameField.typeText(name)
        return self
    }

    @discardableResult
    func enterEmail(_ email: String) -> SignupPage {
        emailField.tap()
        emailField.typeText(email)
        return self
    }

    @discardableResult
    func enterPhone(_ phone: String) -> SignupPage {
        phoneField.tap()
        phoneField.typeText(phone)
        return self
    }

    @discardableResult
    func enterPassword(_ password: String) -> SignupPage {
        passwordField.tap()
        passwordField.typeText(password)
        return self
    }

    @discardableResult
    func enterConfirmPassword(_ password: String) -> SignupPage {
        confirmPasswordField.tap()
        confirmPasswordField.typeText(password)
        return self
    }

    @discardableResult
    func tapSignup() -> SignupPage {
        // No need to dismiss keyboard - tapping button will dismiss it automatically
        signupButton.tap()
        return self
    }

    @discardableResult
    func tapLoginLink() -> LoginPage {
        loginLink.tap()
        return LoginPage(app: app)
    }

    func signup(name: String, email: String, password: String, phone: String? = nil) {
        enterName(name)
        enterEmail(email)
        if let phone = phone {
            enterPhone(phone)
        }
        enterPassword(password)
        tapSignup()
    }

    // MARK: - Assertions

    @discardableResult
    func assertScreenDisplayed(timeout: TimeInterval = 2) -> SignupPage {
        XCTAssertTrue(signupScreen.waitForExistence(timeout: timeout), "Signup screen should be displayed")
        return self
    }

    @discardableResult
    func assertAllElementsVisible() -> SignupPage {
        XCTAssertTrue(nameField.waitForExistence(timeout: 2), "Name field should be visible")
        XCTAssertTrue(emailField.waitForExistence(timeout: 2), "Email field should be visible")
        XCTAssertTrue(passwordField.waitForExistence(timeout: 2), "Password field should be visible")
        XCTAssertTrue(signupButton.waitForExistence(timeout: 2), "Signup button should be visible")
        return self
    }

    @discardableResult
    func assertErrorDisplayed(timeout: TimeInterval = 3) -> SignupPage {
        XCTAssertTrue(errorMessage.waitForExistence(timeout: timeout), "Error message should be displayed")
        return self
    }

    @discardableResult
    func assertValidationDisplayed(timeout: TimeInterval = 2) -> SignupPage {
        XCTAssertTrue(validationMessage.waitForExistence(timeout: timeout), "Validation message should be displayed")
        return self
    }

    // MARK: - Helpers

    private func dismissKeyboardIfNeeded() {
        if app.keyboards.count > 0 {
            app.keyboards.buttons["Return"].tap()
        }
    }
}
